### PR TITLE
Fix material date picker behavior when changing year

### LINKED ISSUE #81547: Unexpected showDatePicker behavior when changing year
It's not clear if this is a bug or intended behavior, but at any rate the Flutter Material date picker has a behavior that differs from that of the native Android date picker in such a way that is likely to cause users accustomed to the native behavior to select the wrong date in a Flutter app.

## Steps to Reproduce

1.  Open a date picker with `showDatePicker` with `initialDate: DateTime(2021, 4, 30)`
2. Tap the year control and select 2015 (or any other year)

**Expected results:** 

4/30/2015 ought to be the newly selected date i.e. the month/day in the newly selected year that corresponds to that of the previously selected date (or closest available). 

**Actual results:** 

The date picker displays April in 2015, but 4/30/2021 remains the selected date. If the user taps OK at this point intending to select 4/30/2015 then they will actually select 4/30/2021.

https://user-images.githubusercontent.com/824653/116717090-8373c600-a9a6-11eb-8528-a24e2849f3de.MP4

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel stable, 2.0.3, on Mac OS X 10.15.7 19H1030 darwin-x64, locale en-US)
    • Flutter version 2.0.3 at /Users/jonathan.joelson/repos/flutter
    • Framework revision 4d7946a68d (6 weeks ago), 2021-03-18 17:24:33 -0700
    • Engine revision 3459eb2436
    • Dart version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/jonathan.joelson/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0
```

</details>

### PR DESCRIPTION
**Pull request by Lexycon**
This PR changes the material date picker behavior when changing the year so that it matches the native picker and the material component guideline. (#81547)

See material component guideline for the date picker: [Material component date-picker behavior](https://m3.material.io/components/date-pickers/guidelines#1531a81f-4052-4a75-a20d-228c7e110156)
See also: [Material components android discussion](https://github.com/material-components/material-components-android/issues/1723)

When selecting another year in the native picker, the same day will be selected (by respecting the boundaries of the date picker). The current material date picker does not select any day when changing the year. This will lead to confusion if the user presses OK and the year does not get updated.

So here is my suggestion:
It will try to preselect the day like the native picker:
 - respecting the boundaries of the date picker (firstDate, lastDate)
 - changing from leapyear 29th february will set 28th february if not a leapyear is selected
 - only set the day if it is selectable (selectableDayPredicate)

The calendar shown in the recording was setup with this parameters:
```
firstDate: DateTime(2016, DateTime.june, 9),
initialDate: DateTime(2018, DateTime.may, 4),
lastDate: DateTime(2021, DateTime.january, 15),
```
 
https://github.com/flutter/flutter/assets/13588771/3041c296-b9d0-4078-88cd-d1135fc343b3

Fixes #81547 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by HansMuller — APPROVED** [review]
Looks like a good fix. @TahaTesser - please take a look

File: packages/flutter/lib/src/material/calendar_date_picker.dart [hunk 1]
```diff
@@ -237,6 +237,10 @@ class _CalendarDatePickerState extends State<CalendarDatePicker> {
   void _handleYearChanged(DateTime value) {
     _vibrate();
 
+    final int daysInMonth = DateUtils.getDaysInMonth(value.year, value.month);
+    final int preferredDay = math.min(_selectedDate.day, daysInMonth);
+    value = DateTime(value.year, value.month, preferredDay);
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/calendar_date_picker.dart:242]:
```suggestion
    value = value.copyWith(day: preferredDay);
```

File: packages/flutter/lib/src/material/calendar_date_picker.dart [hunk 2]
```diff
@@ -257,6 +266,11 @@ class _CalendarDatePickerState extends State<CalendarDatePicker> {
     });
   }
 
+  bool _isSelectable(DateTime date) {
+    return widget.selectableDayPredicate == null ||
+        widget.selectableDayPredicate!.call(date);
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/calendar_date_picker.dart:271]:
```suggestion
    return widget.selectableDayPredicate == null || widget.selectableDayPredicate!.call(date);
```

**Review by TahaTesser — CHANGES_REQUESTED** [review]
Minor nits

**Review by TahaTesser — APPROVED** [review]
LGTM

**Pull request merged**